### ISSUE TITLE
Namespaces: Allow the same task and service name across modules

### ISSUE THREAD

**Issue by edvald** [garden-io/garden#1177]
## Problem

Currently, we have a single namespace for all tasks and services. While this is a good default in most cases, there are other cases where it would be helpful if identifiers could be duplicated across modules.

For example, each of your data storage modules might have a `cleanup` task, and it's inconvenient to have to prefix the name of each of those tasks.

This is also inconsistent with _tests_ which are namespaced per individual module.

## Suggestion

Instead of disallowing duplicate service and task names across modules, we should instead disallow _ambiguous references_. For example, we allow multiple modules to contain a `cleanup` task, but we throw an error if another service or task then lists `cleanup` as a dependency. Instead, the user then needs to specify `<module-name>.cleanup`. Same goes for CLI/API arguments that specify services or tasks.

We should furthermore allow service names and task names to overlap within modules. Same rule applies, we throw on ambiguous references, but this time we'd need to disambiguate by specifying `task:<name>` or `service:<name>`.

We should make sure that the errors we throw are helpful, and provide the list of possible references to disambiguate with.

The above has the strong benefit of being backwards-compatible with the current setup, and also retains sane default behavior for when you don't need this flexibility.

## Future work

We can build on this to allow more flexibility in dependencies, e.g. allowing builds to depend on tasks and tests.

We should also consider that we may add yet another namespace layer, when we start properly supporting sub-projects (so we might have references like `task:some-project/some-module.some-task`). 

Open question: How do we feel about that syntax?

**Comment by burdiyan** [comment]
+1 on this one. Also Garden fails weirdly when having multiple tasks with the same name:

```

[2019-11-14T13:13:13.102Z] TypeError: Cannot read property 'match' of undefined
    at Object.getModuleKey (/snapshot/dist/build/src/types/module.js:0)
    at Garden.<anonymous> (/snapshot/dist/build/src/garden.js:0)
    at Generator.next (<anonymous>)
    at __awaiter (/snapshot/dist/build/src/garden.js:0)
    at new Promise (<anonymous>)
    at __awaiter (/snapshot/dist/build/src/garden.js:0)
    at Garden.addModule (/snapshot/dist/build/src/garden.js:0)
    at Garden.<anonymous> (/snapshot/dist/build/src/garden.js:0)
    at Generator.next (<anonymous>)
    at __awaiter (/snapshot/dist/build/src/garden.js:0)
    at new Promise (<anonymous>)
    at __awaiter (/snapshot/dist/build/src/garden.js:0)
    at scanModules.__awaiter.asyncLock.acquire.bluebird_1.default.map (/snapshot/dist/build/src/garden.js:0)
    at tryCatcher (/snapshot/dist/node_modules/bluebird/js/release/util.js:16:23)
    at MappingPromiseArray.module.exports.MappingPromiseArray._promiseFulfilled (/snapshot/dist/node_modules/bluebird/js/release/map.js:61:38)
    at MappingPromiseArray.module.exports.PromiseArray._iterate (/snapshot/dist/node_modules/bluebird/js/release/promise_array.js:114:31)
    at MappingPromiseArray.init (/snapshot/dist/node_modules/bluebird/js/release/promise_array.js:78:10)
    at MappingPromiseArray.module.exports.MappingPromiseArray._asyncInit (/snapshot/dist/node_modules/bluebird/js/release/map.js:30:10)
    at _drainQueueStep (/snapshot/dist/node_modules/bluebird/js/release/async.js:142:12)
    at _drainQueue (/snapshot/dist/node_modules/bluebird/js/release/async.js:131:9)
    at Async._drainQueues (/snapshot/dist/node_modules/bluebird/js/release/async.js:147:5)
    at Immediate.Async.drainQueues (/snapshot/dist/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
```

**Comment by edvald** [comment]
@burdiyan that bug may be unrelated, but we should look into that. Could you file it separately, along with some context? (e.g. the module name(s) and task names in question)
